internal/k8s/replicaset/replicaset.go

```diff
@@ -0,0 +1,152 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+package replicaset
+
+import (
+	"context"
+	"reflect"
+	"strings"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/vdaas/vald/internal/errors"
+	"github.com/vdaas/vald/internal/k8s"
+	"github.com/vdaas/vald/internal/log"
+)
+
+// ReplicaSetWatcher is a type alias for k8s resource controller.
+type ReplicaSetWatcher k8s.ResourceController
+
+type reconciler struct {
+	ctx                  context.Context
+	mgr                  manager.Manager
+	name                 string
+	namespace            string
+	onError              func(err error)
+	onReconcile          func(rs map[string][]ReplicaSet)
+	lastReconciledResult map[string][]ReplicaSet
+}
+
+// ReplicaSet is a type alias for the k8s replica set definition.
+type ReplicaSet = appsv1.ReplicaSet
+
+// New returns the ReplicaSetWatcher that implements reconciliation loop, or any error occurred.
+func New(opts ...Option) (ReplicaSetWatcher, error) {
+	r := &reconciler{
+		lastReconciledResult: make(map[string][]ReplicaSet),
+	}
+
+	for _, opt := range opts {
+		if err := opt(r); err != nil {
+			return nil, errors.ErrOptionFailed(err, reflect.ValueOf(opt))
+		}
+	}
+
+	return r, nil
+}
+
+// Reconcile implements k8s reconciliation loop to retrieve the ReplicaSet information from k8s.
+func (r *reconciler) Reconcile(req reconcile.Request) (res reconcile.Result, err error) {
+	rsl := new(appsv1.ReplicaSetList)
+
+	err = r.mgr.GetClient().List(r.ctx, rsl)
+	if err != nil {
+		if r.onError != nil {
+			r.onError(err)
+		}
+		res = reconcile.Result{
+			Requeue:      true,
+			RequeueAfter: time.Millisecond * 100,
+		}
+		if k8serrors.IsNotFound(err) {
+			log.Warn(errors.ErrK8sResourceNotFound(err))
+			res.RequeueAfter = time.Second
+			return res, nil
+		}
+		return
+	}
+
+	rs := make(map[string][]ReplicaSet)
+
+	for _, replicaset := range rsl.Items {
+		name, ok := replicaset.GetObjectMeta().GetLabels()["app"]
+		if !ok {
+			pns := strings.Split(replicaset.GetName(), "-")
+			name = strings.Join(pns[:len(pns)-1], "-")
+		}
+
+		if _, ok := rs[name]; !ok {
+			lrr, ok := r.lastReconciledResult[name]
+			if !ok {
+				rs[name] = make([]ReplicaSet, 0, 0)
+			} else {
+				rs[name] = make([]ReplicaSet, 0, len(lrr))
+			}
+		}
+
+		rs[name] = append(rs[name], replicaset)
+	}
+
+	r.lastReconciledResult = rs
+
+	if r.onReconcile != nil {
+		r.onReconcile(rs)
```

Comment:
how do you think about race condition, if onReconcile function includes goroutine which makes race condition problem